### PR TITLE
Fixups + stats for partitioning columns

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java [hunk 1]
```diff
@@ -146,7 +146,7 @@ default boolean canCreateSampledTables(ConnectorSession session)
     /**
      * Get statistics for table using given layout.
      */
-    default TableStatistics getTableStatistics(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorTableLayoutHandle layoutHandle)
+    default TableStatistics getTableStatistics(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorTableHandle tableHandle, ConnectorTableLayoutHandle layoutHandle)
```
Comment (ArturGajowy) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java]: BTW: why is transaction needed for fetching statistics? Same for session?

Comment (ArturGajowy) [reply presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java]: why is layout not enough? Doesn't a layout _belong to_ a table?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 2]
```diff
@@ -469,13 +469,9 @@ private Estimate getColumnStatisticsValue(
...
 
-    private Map<String, ColumnHandle> getTableColumnNames(ConnectorTableLayoutHandle layoutHandle, ConnectorSession session)
+    private Map<String, ColumnHandle> getTableColumnNames(List<HivePartition> partitions, ConnectorSession session)
     {
-        HiveTableLayoutHandle layout = checkType(layoutHandle, HiveTableLayoutHandle.class, "layoutHandle");
-        if (!layout.getPartitions().isPresent()) {
-            return ImmutableMap.of();
-        }
-        HivePartition hivePartition = Iterables.getFirst(layout.getPartitions().get(), null);
+        HivePartition hivePartition = Iterables.getFirst(partitions, null);
```
Comment (ArturGajowy) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: static import onegai

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 3]
```diff
@@ -469,13 +469,9 @@ private Estimate getColumnStatisticsValue(
         }
     }
 
-    private Map<String, ColumnHandle> getTableColumnNames(ConnectorTableLayoutHandle layoutHandle, ConnectorSession session)
+    private Map<String, ColumnHandle> getTableColumnNames(List<HivePartition> partitions, ConnectorSession session)
```
Comment (ArturGajowy) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: `getColumnHandlesByTableName`?

Comment (ArturGajowy) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: `...forTableNames`?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 4]
```diff
@@ -382,8 +382,14 @@ public TableStatistics getTableStatistics(ConnectorTransactionHandle transaction
             String columnName = columnEntry.getKey();
             HiveColumnHandle hiveColumnHandle = (HiveColumnHandle) columnEntry.getValue();
             com.facebook.presto.spi.statistics.ColumnStatistics.Builder columnStatistics = com.facebook.presto.spi.statistics.ColumnStatistics.builder();
```
Comment (ArturGajowy) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: missing import?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 5]
```diff
@@ -445,6 +456,36 @@ private Estimate calculateNullsCount(Map<String, PartitionStatistics> partitionS
...
 
+    private Estimate calculateDistinctValuesCountForPartitioningKey(HiveColumnHandle hiveColumnHandle, List<HivePartition> partitions)
+    {
+        return new Estimate(partitions.stream()
+                .map(partition -> partition.getKeys().get(hiveColumnHandle))
+                .collect(toSet())
+                .size());
+    }
+
+    private Estimate calculateNullsCountForPartitioningKey(HiveColumnHandle hiveColumnHandle, List<HivePartition> hivePartitions, Map<String, PartitionStatistics> partitionStatisticsMap)
```
Comment (ArturGajowy) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:467]: `partitionStatisticsMap` -> `statisticsByPartitionName`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 6]
```diff
@@ -379,27 +379,38 @@ public TableStatistics getTableStatistics(ConnectorTransactionHandle transaction
         tableStatistics.setRowCount(calculateRowsCount(partitionStatisticsMap));
 
         for (Map.Entry<String, ColumnHandle> columnEntry : tableColumns.entrySet()) {
+            String columnName = columnEntry.getKey();
+            HiveColumnHandle hiveColumnHandle = (HiveColumnHandle) columnEntry.getValue();
             com.facebook.presto.spi.statistics.ColumnStatistics.Builder columnStatistics = com.facebook.presto.spi.statistics.ColumnStatistics.builder();
-            columnStatistics.setDistinctValuesCount(calculateDistinctValuesCount(partitionStatisticsMap, columnEntry.getKey()));
-            columnStatistics.setNullsCount(calculateNullsCount(partitionStatisticsMap, columnEntry.getKey()));
-            tableStatistics.setColumnStatistics(columnEntry.getValue(), columnStatistics.build());
+            if (hiveColumnHandle.isPartitionKey()) {
```
Comment (ArturGajowy) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:385]: I'm wondering if we could have a strategy for computing null and unique value counts (two separate methods) with two implemenatations (for columns that are partition keys and for other columns)?

Alternarively, a strategy with a single method computing a `ColumnSummary {nullsCount, uniqueValuesCount }`? (one method; two strategy implementations)

**Comment by ArturGajowy** [comment]
Please do the renames and the import, the rest of the comments are more of optional / discussion. Ohterwise LGTM.

**Comment by losipiuk** [comment]
I did what I could.
It would be nice to a history rewrite to `statistics6`. There is a lot of fixup commits. It is hard to maintain that.

**Comment by ArturGajowy** [comment]
Let's have the `statistics6` branch, we can rebase show_column_stats on it.

**Pull request closed without merging**